File: java/fizzbuzz-kafka-streams/src/main/kotlin/FizzBuzz.kt
import arrow.core.Option
import arrow.core.getOrElse

class FizzBuzz {
    private val rules = listOf(
            Rule(3, "Fizz"),
            Rule(5, "Buzz")
    )

    fun sequenceUpTo(limit: Int) = (1..limit).map(::single)

    fun single(input: Int) = Option.just(tryGetResultFromRules(input))
            .filter { it.isNotEmpty() }
            .getOrElse { input.toString() }

    private fun tryGetResultFromRules(input: Int) = rules
            .asSequence()
            .filter { it.appliesTo(input) }
            .joinToString("-") { it.result }
}

data class Rule(private val divisor: Int, val result: String) {
    fun appliesTo(input: Int) = input % divisor == 0
}

File: svelte/fizz-buzz/src/fizzbuzz.ts
const rules = [
  {divisor: 3, result: 'Fizz'},
  {divisor: 5, result: 'Buzz'},
];

export const fizzBuzz = (input: number): string =>
  rules.filter(rule => input % rule.divisor === 0)
       .map(rule => rule.result)
       .join('-') || input.toString();
